[Bug]: Prometheus metrics storage panics at startup when latency_unit is set to an invalid value

### What happened?

Setting `latency_unit` to anything other than `ms` or `s` in the v2 YAML config passes validation silently, then panics inside `NewMetricsReader` at startup with `"programming error: unknown latency unit"`.

The v1 flag path already catches this: `--prometheus.query.duration-unit` returns a clean error. The v2 YAML path skips that check entirely because `LatencyUnit` has no validation tag in `promcfg.Configuration`.

Leaving it unset is fine (defaults to `ms`). The panic only hits when you explicitly set a wrong value with `normalize_duration: true`. Confirmed locally: `Validate()` returns nil on `latency_unit: us`, then `NewMetricsReader` panics.

### Steps to reproduce

1. Start from a Jaeger v2 SPM config with a Prometheus metrics backend (for example `cmd/jaeger/config-spm.yaml`).
2. Under the prometheus backend, set:
   ```yaml
   normalize_duration: true
   latency_unit: us
   ```
3. Start Jaeger. Config validation passes.
4. Jaeger panics while building the metrics reader and exits.

code to repro issue : 
```go
// run from within the jaeger repo (imports internal packages)
package main

import (
    promcfg "github.com/jaegertracing/jaeger/internal/config/promcfg"
    prom "github.com/jaegertracing/jaeger/internal/storage/metricstore/prometheus/metricstore"
    "go.opentelemetry.io/otel/trace/noop"
    "go.uber.org/zap"
)

func main() {
    cfg := promcfg.Configuration{
        ServerURL:         "http://localhost:9090",
        NormalizeDuration: true,
        LatencyUnit:       "us", // invalid: only ms or s are supported
    }
    _ = cfg.Validate()
    _, _ = prom.NewMetricsReader(cfg, zap.NewNop(), noop.NewTracerProvider(), nil)
}
```

### Expected behavior

Config validation should reject the bad value with a clear error, like the v1 flag path does. A typo in the config
should not cause a panic.

### Relevant log output

```shell
panic: programming error: unknown latency unit: us
```

### Screenshot

<img width="1310" height="641" alt="Image" src="https://github.com/user-attachments/assets/7f4b494c-f572-4d7d-bbb1-2e02b803cd7e" />

### Additional context

- `internal/config/promcfg/config.go`: `LatencyUnit` has no `valid:` tag, and `Validate()` only calls govalidator.
- `internal/storage/metricstore/prometheus/metricstore/reader.go`: `buildFullLatencyMetricName` panics on an unknown unit inside `NewMetricsReader`.
- `internal/storage/metricstore/prometheus/options.go`: the v1 path that already validates this. The same check can be reused.

A simple fix is to check `latency_unit` in `Validate()` (allow empty since the default is applied later).

cc @yurishkuro 

### Jaeger backend version

Jaeger v2 (current main)

### SDK

n/a, this is a config issue

### Pipeline

OTEL SDK -> Jaeger v2 (SPM) -> Prometheus

### Stogage backend

Prometheus

### Operating system

Any

### Deployment model

Any

### Deployment configs

```shell
prometheus:
  endpoint: http://localhost:9090
  normalize_duration: true
  latency_unit: us
```

--- a/internal/config/promcfg/config.go
+++ b/internal/config/promcfg/config.go
@@ -34,16 +34,16 @@ func (c *Configuration) Validate() error {
 	if _, err := govalidator.ValidateStruct(c); err != nil {
 		return err
 	}
-	// LatencyUnit must be "ms" or "s". An empty value is allowed because the
-	// default ("ms") is applied during config unmarshalling. Without this check
-	// an invalid unit passes validation and later panics when the metric name
-	// is built (see buildFullLatencyMetricName). This mirrors the validation
-	// already done for the v1 flag path in
-	// internal/storage/metricstore/prometheus/options.go.
-	switch c.LatencyUnit {
-	case "", "ms", "s":
-	default:
+	// Empty is allowed because the default ("ms") is applied during unmarshalling.
+	if c.LatencyUnit != "" && !IsValidLatencyUnit(c.LatencyUnit) {
 		return fmt.Errorf(`latency_unit must be "ms" or "s", not %q`, c.LatencyUnit)
 	}
 	return nil
+}
+
+// IsValidLatencyUnit reports whether u is a latency unit the Prometheus metric
+// name builder understands. It is the single source of truth shared by this
+// validation and the v1 flag path in prometheus/options.go.
+func IsValidLatencyUnit(u string) bool {
+	return u == "ms" || u == "s"
 }

--- a/internal/config/promcfg/config.go
+++ b/internal/config/promcfg/config.go
@@ -4,6 +4,7 @@
 package promcfg
 
 import (
+	"fmt"
 	"time"
 
 	"github.com/asaskevich/govalidator"
@@ -30,6 +31,19 @@ type Configuration struct {
 }
 
 func (c *Configuration) Validate() error {
-	_, err := govalidator.ValidateStruct(c)
-	return err
+	if _, err := govalidator.ValidateStruct(c); err != nil {
+		return err
+	}
+	// LatencyUnit must be "ms" or "s". An empty value is allowed because the
+	// default ("ms") is applied during config unmarshalling. Without this check
+	// an invalid unit passes validation and later panics when the metric name
+	// is built (see buildFullLatencyMetricName). This mirrors the validation
+	// already done for the v1 flag path in
+	// internal/storage/metricstore/prometheus/options.go.
+	switch c.LatencyUnit {
+	case "", "ms", "s":
+	default:
+		return fmt.Errorf(`latency_unit must be "ms" or "s", not %q`, c.LatencyUnit)
+	}
+	return nil
 }

--- a/internal/config/promcfg/config.go
+++ b/internal/config/promcfg/config.go
@@ -34,9 +34,13 @@ func (c *Configuration) Validate() error {
 	if _, err := govalidator.ValidateStruct(c); err != nil {
 		return err
 	}
-	// Empty is allowed because the default ("ms") is applied during unmarshalling.
+	// An empty LatencyUnit is allowed: it means "use the default" ("ms"). That
+	// default is normally populated when the config is loaded (the CLI flag
+	// default for v1, DefaultConfig for v2), but programmatically-constructed
+	// configs may leave it empty, so callers must not assume LatencyUnit is
+	// non-empty after Validate.
 	if c.LatencyUnit != "" && !IsValidLatencyUnit(c.LatencyUnit) {
-		return fmt.Errorf(`latency_unit must be "ms" or "s", not %q`, c.LatencyUnit)
+		return LatencyUnitError(c.LatencyUnit)
 	}
 	return nil
 }
@@ -46,4 +50,11 @@ func (c *Configuration) Validate() error {
 // validation and the v1 flag path in prometheus/options.go.
 func IsValidLatencyUnit(u string) bool {
 	return u == "ms" || u == "s"
+}
+
+// LatencyUnitError reports that the configured latency unit is unsupported. It is
+// shared by the v2 config validation and the v1 flag path so both surfaces emit
+// an identical message.
+func LatencyUnitError(value string) error {
+	return fmt.Errorf(`latency_unit must be "ms" or "s", not %q`, value)
 }

--- a/internal/config/promcfg/config_test.go
+++ b/internal/config/promcfg/config_test.go
@@ -22,6 +22,34 @@ func TestValidate(t *testing.T) {
 	require.Error(t, err)
 }
 
+func TestValidateLatencyUnit(t *testing.T) {
+	tests := []struct {
+		name        string
+		latencyUnit string
+		wantErr     string
+	}{
+		{name: "empty is allowed (default applied later)", latencyUnit: ""},
+		{name: "ms is valid", latencyUnit: "ms"},
+		{name: "s is valid", latencyUnit: "s"},
+		{name: "invalid unit is rejected", latencyUnit: "us", wantErr: "latency_unit"},
+		{name: "long form is rejected", latencyUnit: "milliseconds", wantErr: "latency_unit"},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			cfg := Configuration{
+				ServerURL:   "localhost:1234",
+				LatencyUnit: test.latencyUnit,
+			}
+			err := cfg.Validate()
+			if test.wantErr != "" {
+				require.ErrorContains(t, err, test.wantErr)
+				return
+			}
+			require.NoError(t, err)
+		})
+	}
+}
+
 func TestMain(m *testing.M) {
 	testutils.VerifyGoLeaks(m)
 }

--- a/internal/config/promcfg/config_test.go
+++ b/internal/config/promcfg/config_test.go
@@ -33,11 +33,12 @@ func TestValidateLatencyUnit(t *testing.T) {
 		{name: "s is valid", latencyUnit: "s"},
 		{name: "invalid unit is rejected", latencyUnit: "us", wantErr: "latency_unit"},
 		{name: "long form is rejected", latencyUnit: "milliseconds", wantErr: "latency_unit"},
+		{name: "uppercase is rejected (case-sensitive)", latencyUnit: "MS", wantErr: "latency_unit"},
 	}
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
 			cfg := Configuration{
-				ServerURL:   "localhost:1234",
+				ServerURL:   "http://localhost:1234",
 				LatencyUnit: test.latencyUnit,
 			}
 			err := cfg.Validate()

--- a/internal/storage/metricstore/prometheus/factory.go
+++ b/internal/storage/metricstore/prometheus/factory.go
@@ -56,9 +56,9 @@ func NewFactoryWithConfig(
 		return nil, err
 	}
 	// Validate permits an empty LatencyUnit (it means "use the default"), so
-	// normalize it here before storing. Otherwise a config with NormalizeDuration
-	// set would reach the reader with an empty unit and panic when building the
-	// metric name.
+	// normalize it to the default before storing so downstream code never sees an
+	// empty unit. Without this, a config with NormalizeDuration enabled would
+	// reach the reader with an empty unit and panic when building the metric name.
 	if cfg.LatencyUnit == "" {
 		cfg.LatencyUnit = defaultLatencyUnit
 	}

--- a/internal/storage/metricstore/prometheus/factory.go
+++ b/internal/storage/metricstore/prometheus/factory.go
@@ -55,6 +55,13 @@ func NewFactoryWithConfig(
 	if err := cfg.Validate(); err != nil {
 		return nil, err
 	}
+	// Validate permits an empty LatencyUnit (it means "use the default"), so
+	// normalize it here before storing. Otherwise a config with NormalizeDuration
+	// set would reach the reader with an empty unit and panic when building the
+	// metric name.
+	if cfg.LatencyUnit == "" {
+		cfg.LatencyUnit = defaultLatencyUnit
+	}
 	f := NewFactory()
 	f.options = &Options{
 		Configuration: cfg,

--- a/internal/storage/metricstore/prometheus/factory_test.go
+++ b/internal/storage/metricstore/prometheus/factory_test.go
@@ -95,12 +95,11 @@ func TestWithConfiguration(t *testing.T) {
 			ServerURL:   "http://localhost:9090",
 			LatencyUnit: "milliseconds",
 		}
-		// NewFactoryWithConfig should validate and reject invalid latency unit
-		// However, the validation is currently not implemented in Configuration.Validate()
-		// So this test now just creates the factory successfully
-		f, err := NewFactoryWithConfig(cfg, telemetry.NoopSettings(), nil)
-		require.NoError(t, err)
-		assert.Equal(t, "milliseconds", f.options.LatencyUnit)
+		// NewFactoryWithConfig validates the config and rejects an invalid
+		// latency unit instead of letting it panic later when the metric name
+		// is built.
+		_, err := NewFactoryWithConfig(cfg, telemetry.NoopSettings(), nil)
+		require.ErrorContains(t, err, "latency_unit")
 	})
 }
 

--- a/internal/storage/metricstore/prometheus/factory_test.go
+++ b/internal/storage/metricstore/prometheus/factory_test.go
@@ -99,7 +99,7 @@ func TestWithConfiguration(t *testing.T) {
 		// latency unit instead of letting it panic later when the metric name
 		// is built.
 		_, err := NewFactoryWithConfig(cfg, telemetry.NoopSettings(), nil)
-		require.ErrorContains(t, err, "latency_unit")
+		require.EqualError(t, err, `latency_unit must be "ms" or "s", not "milliseconds"`)
 	})
 }
 

--- a/internal/storage/metricstore/prometheus/factory_test.go
+++ b/internal/storage/metricstore/prometheus/factory_test.go
@@ -101,6 +101,22 @@ func TestWithConfiguration(t *testing.T) {
 		_, err := NewFactoryWithConfig(cfg, telemetry.NoopSettings(), nil)
 		require.EqualError(t, err, `latency_unit must be "ms" or "s", not "milliseconds"`)
 	})
+	t.Run("empty latency unit with normalize_duration is normalized to the default", func(t *testing.T) {
+		cfg := promcfg.Configuration{
+			ServerURL:         "http://localhost:9090",
+			NormalizeDuration: true,
+			LatencyUnit:       "", // empty: must be normalized to the default, not panic later
+		}
+		f, err := NewFactoryWithConfig(cfg, telemetry.NoopSettings(), nil)
+		require.NoError(t, err)
+		assert.Equal(t, "ms", f.options.LatencyUnit)
+		// Building the reader must not panic now that the unit is normalized.
+		require.NotPanics(t, func() {
+			reader, err := f.CreateMetricsReader()
+			require.NoError(t, err)
+			assert.NotNil(t, reader)
+		})
+	})
 }
 
 func TestEmptyFactoryConfig(t *testing.T) {

--- a/internal/storage/metricstore/prometheus/options.go
+++ b/internal/storage/metricstore/prometheus/options.go
@@ -119,8 +119,7 @@ func (opt *Options) InitFromViper(v *viper.Viper) error {
 		return fmt.Errorf("failed to parse extra query params: %w", err)
 	}
 
-	isValidUnit := map[string]bool{"ms": true, "s": true}
-	if _, ok := isValidUnit[opt.LatencyUnit]; !ok {
+	if !config.IsValidLatencyUnit(opt.LatencyUnit) {
 		return fmt.Errorf(`duration-unit must be one of "ms" or "s", not %q`, opt.LatencyUnit)
 	}
 

--- a/internal/storage/metricstore/prometheus/options.go
+++ b/internal/storage/metricstore/prometheus/options.go
@@ -120,7 +120,7 @@ func (opt *Options) InitFromViper(v *viper.Viper) error {
 	}
 
 	if !config.IsValidLatencyUnit(opt.LatencyUnit) {
-		return fmt.Errorf(`duration-unit must be one of "ms" or "s", not %q`, opt.LatencyUnit)
+		return config.LatencyUnitError(opt.LatencyUnit)
 	}
 
 	tlsCfg, err := tlsFlagsCfg.InitFromViper(v)

--- a/internal/storage/metricstore/prometheus/options_test.go
+++ b/internal/storage/metricstore/prometheus/options_test.go
@@ -46,6 +46,15 @@ func TestInitFromViperLatencyUnit(t *testing.T) {
 		require.NoError(t, err)
 		assert.Equal(t, "s", opts.LatencyUnit)
 	})
+	t.Run("empty unit is rejected at the flag layer", func(t *testing.T) {
+		// Unlike NewFactoryWithConfig (which normalizes empty to the default),
+		// the v1 flag path rejects an explicitly empty unit.
+		opts := NewOptions()
+		v, _ := jconfig.Viperize(opts.AddFlags)
+		v.Set(prefix+suffixLatencyUnit, "")
+		err := opts.InitFromViper(v)
+		require.EqualError(t, err, `latency_unit must be "ms" or "s", not ""`)
+	})
 }
 
 func TestParseKV(t *testing.T) {

--- a/internal/storage/metricstore/prometheus/options_test.go
+++ b/internal/storage/metricstore/prometheus/options_test.go
@@ -9,7 +9,9 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 
+	jconfig "github.com/jaegertracing/jaeger/internal/config"
 	config "github.com/jaegertracing/jaeger/internal/config/promcfg"
 )
 
@@ -26,6 +28,24 @@ func TestCLI(t *testing.T) {
 func TestCLIError(t *testing.T) {
 	_, err := parseKV("key1")
 	assert.ErrorContains(t, err, "failed to parse 'key1'. Expected format: 'param1=value1,param2=value2'")
+}
+
+func TestInitFromViperLatencyUnit(t *testing.T) {
+	t.Run("invalid unit is rejected", func(t *testing.T) {
+		opts := NewOptions()
+		v, _ := jconfig.Viperize(opts.AddFlags)
+		v.Set(prefix+suffixLatencyUnit, "us")
+		err := opts.InitFromViper(v)
+		require.EqualError(t, err, `latency_unit must be "ms" or "s", not "us"`)
+	})
+	t.Run("valid unit is accepted", func(t *testing.T) {
+		opts := NewOptions()
+		v, _ := jconfig.Viperize(opts.AddFlags)
+		v.Set(prefix+suffixLatencyUnit, "s")
+		err := opts.InitFromViper(v)
+		require.NoError(t, err)
+		assert.Equal(t, "s", opts.LatencyUnit)
+	})
 }
 
 func TestParseKV(t *testing.T) {

